### PR TITLE
Specify separate directories to run external instances in ExternalGeneratorFilter

### PR DESCRIPTION
**Pull request by colizz**
#### PR description:

In the ExternalGeneratorFilter module, external cmsRun instances are launched by the main instance in the same working directory. 
As some types of generator will release intermediate files (e.g. H7, Sherpa), we execute the commands in different directories (specifed as `thread0`, `thread1`, ...) to avoid overlapping.

#### PR validation:

Validated over all possible GeneratorFilter-based modules as summarised in https://github.com/cms-sw/cmssw/pull/34146#issuecomment-863273408. 
Physics results show good consistency.

### CODE REVIEW

File: GeneratorInterface/Core/plugins/ExternalGeneratorFilter.cc [hunk 1]
```diff
@@ -46,15 +46,22 @@ namespace externalgen {
...
         edm::LogSystem("ExternalProcess") << id_ << " starting external process \n";
         std::string verboseCommand;
         if (verbose) {
           verboseCommand = "--verbose ";
         }
+        std::error_code ec;
+        auto curDir = current_path();
+        auto newDir = path("thread"s + std::to_string(id_));
+        create_directory(newDir, ec);
+        current_path(newDir, ec);
```
Comment (makortel) [inline GeneratorInterface/Core/plugins/ExternalGeneratorFilter.cc:59]: I'm concerned that any possible errors are silently ignored.

Comment (colizz) [reply GeneratorInterface/Core/plugins/ExternalGeneratorFilter.cc:59]: Thanks, shall this work?
```suggestion
        current_path(newDir, ec);
        if (ec) {
          throw cms::Exception("ExternalFailed")
            << "failed entering the working dir for process " << id_ << ": " << ec.message();
        }
```
(Actually the previous code was following this example https://github.com/cms-sw/cmssw/blob/c5d26737f2e7f2a28d2e485b18f07da4d4d79502/GeneratorInterface/LHEInterface/plugins/ExternalLHEProducer.cc#L481-L483)

Comment (makortel) [reply GeneratorInterface/Core/plugins/ExternalGeneratorFilter.cc:59]: How about letting these functions to throw the exceptions? (i.e. remove the `ec` altogether, assuming the messages of those exceptions would be clear-enough)

**Comment by agrohsje** [comment]
please test

**Comment by agrohsje** [comment]
generators

**Pull request merged**